src/daft-recordbatch/src/ops/window.rs

```diff
@@ -93,38 +93,227 @@ impl RecordBatch {
+    fn calculate_row_frame_bounds(
+        row_idx: usize,
+        start_boundary: Option<i64>,
+        end_boundary: Option<i64>,
+        total_rows: usize,
+    ) -> DaftResult<(usize, usize)> {
+        let frame_start = match start_boundary {
+            None => 0, // Unbounded preceding
+            Some(offset) => row_idx
+                .saturating_add_signed(offset as isize)
+                .min(total_rows),
+        };
+
+        let frame_end = match end_boundary {
+            None => total_rows, // Unbounded following
+            Some(offset) => (row_idx + 1)
+                .saturating_add_signed(offset as isize)
+                .min(total_rows),
+        };
+
+        if frame_end < frame_start {
+            return Err(DaftError::ValueError(
+                "Negative frame size is not allowed".into(),
+            ));
+        }
+
+        Ok((frame_start, frame_end))
+    }
+
+    // Helper method to calculate range-based frame boundaries
+    #[allow(clippy::too_many_arguments)]
+    fn calculate_range_frame_bounds(
+        row_idx: usize,
+        prev_frame_start: usize,
+        prev_frame_end: usize,
+        start_boundary: Option<&WindowBoundary>,
+        end_boundary: Option<&WindowBoundary>,
+        order_by_col: &Series,
+        current_row_order_by: Series,
+        descending: bool,
+        total_rows: usize,
+    ) -> DaftResult<(usize, usize)> {
+        let frame_start = match start_boundary {
+            None => 0, // Unbounded preceding
+            Some(WindowBoundary::Offset(offset)) => row_idx
+                .saturating_add_signed(*offset as isize)
+                .min(total_rows),
+            Some(WindowBoundary::RangeOffset(offset)) => {
+                let lower_bound = (current_row_order_by.clone() + offset.to_series())?;
+                let cmp = |i: usize| -> bool {
+                    if descending {
+                        order_by_col
+                            .slice(i, i + 1)
+                            .unwrap()
+                            .lte(&lower_bound)
+                            .unwrap()
+                            .get(0)
+                            .unwrap_or(false)
+                    } else {
+                        order_by_col
+                            .slice(i, i + 1)
+                            .unwrap()
+                            .gte(&lower_bound)
+                            .unwrap()
+                            .get(0)
+                            .unwrap_or(false)
+                    }
+                };
+                let mut idx = prev_frame_start;
+                while idx < total_rows {
+                    if cmp(idx) {
+                        break;
+                    }
+                    idx += 1;
+                }
+                idx
+            }
+            Some(WindowBoundary::UnboundedPreceding) | Some(WindowBoundary::UnboundedFollowing) => {
+                unreachable!()
+            }
```

Comment:
At this point it might be cleaner and more understandable just to pass in the `WindowBoundary` instead of an `Option<WindowBoundary>`. Alternatively, you can create a very lightweight enum with variants of either unbounded, offset, or range offset.